prow/plugins/commands/policies/policies.go

```diff
@@ -0,0 +1,124 @@
+limitations under the License.
+*/
+
+package policies
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/plugins"
+	"k8s.io/test-infra/prow/plugins/commands"
+)
+
+type AccessPolicy interface {
+	CanAccess(*commands.Context) (bool, error)
+	Who(config *plugins.Configuration) string
+}
+
+func Apply(policy AccessPolicy, handler commands.MatchHandler) commands.MatchHandler {
+	return func(ctx *commands.Context) error {
+		approved, err := policy.CanAccess(ctx)
+		if err != nil {
+			return err
+		} else if !approved {
+			// Commenter is not allowed to use the command.
+			e := ctx.Event
+			msg := fmt.Sprintf("Only %s can use this command.", policy.Who(ctx.Client.PluginConfig))
+			return ctx.Client.GitHubClient.CreateComment(
+				e.Repo.Owner.Login,
+				e.Repo.Name,
+				e.Number,
+				plugins.FormatResponseRaw(e.Body, e.HTMLURL, e.User.Login, msg),
+			)
+		}
+		return handler(ctx)
+	}
+}
+
+type union []AccessPolicy
+
+func Union(options ...AccessPolicy) AccessPolicy {
+	return union(options)
+}
+
+func (a union) CanAccess(ctx *commands.Context) (bool, error) {
+	for _, policy := range a {
+		if allowed, err := policy.CanAccess(ctx); err != nil {
+			return false, err
+		} else if allowed {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (a union) Who(config *plugins.Configuration) string {
+	strs := make([]string, 0, len(a))
+	for _, policy := range a {
+		strs = append(strs, policy.Who(config))
+	}
+	return fmt.Sprintf("[%s]", strings.Join(strs, " AND "))
+}
+
+// Possible policies: any, org member, org member + assign, GH team, issue/pr author, OWNERS alias, user whitelist...
+
+var _ = AccessPolicy(OpenAccess{})
+var _ = AccessPolicy(TeamAccess(nil))
+
+// OpenAccess allows anyone to use a command.
+type OpenAccess struct{}
+
+func (o OpenAccess) CanAccess(*commands.Context) (bool, error) {
+	return true, nil
+}
+
+func (o OpenAccess) Who(*plugins.Configuration) string {
+	return "Everyone"
+}
+
+// teamAccess allows members of a specific GitHub team to use a command.
+type teamAccess func(*plugins.Configuration) (teamName string, teamID int)
+
+func TeamAccess(teamGetter func(*plugins.Configuration) (teamName string, teamID int)) AccessPolicy {
+	return teamAccess(teamGetter)
+}
+
+func (t teamAccess) CanAccess(ctx *commands.Context) (bool, error) {
+	name, id := t(ctx.Client.PluginConfig)
+	members, err := ctx.Client.GitHubClient.ListTeamMembers(id)
+	if err != nil {
+		return false, fmt.Errorf("error listing members of team %q (ID: %d): %v", name, id, err)
+	}
+
+	found := false
+	for _, person := range members {
+		login := github.NormLogin(ctx.Event.User.Login)
+		if github.NormLogin(person.Login) == login {
+			found = true
+			break
+		}
+	}
+	return found, nil
+}
+
+func (t teamAccess) Who(config *plugins.Configuration) string {
+	name, _ := t(config)
+	return fmt.Sprintf("Members of the %q GitHub team", name)
```

Comment:
nit: if this is being string formatted into `"Only %s can` we should lowercase the first word of `Who()`

Reply:
It is also used in a different format string in generics.go.  I changed the `Only %s can` format string to be appropriate for capitalized values.